### PR TITLE
[17.0] [MIG] account_payment_term_extension

### PR DESCRIPTION
**Pull request by AnaisLForgeFlow**
Standard migration

@ForgeFlow

### CODE REVIEW

File: account_payment_term_extension/models/account_payment_term_line.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+                ]
+            if payment_terms_delay_type == "weeks_and_months":
+                date_type += [
+                    ("weeks_after", "Weeks after invoice date"),
+                    ("weeks_after_end_of_month", "Weeks after end of month"),
+                    ("weeks_after_end_of_next_month", "Weeks after end of next month"),
+                    ("months_after", "Months after invoice date"),
+                    ("months_after_end_of_month", "Months after end of month"),
+                ]
+        return date_type
```
Comment (GuillemCForgeFlow) [inline account_payment_term_extension/models/account_payment_term_line.py]:
```suggestion
        weeks_delay_type = [
            ("weeks_after", "Weeks after invoice date"),
            ("weeks_after_end_of_month", "Weeks after end of month"),
            ("weeks_after_end_of_next_month", "Weeks after end of next month"),
        ]
        months_delay_type = [
            ("months_after", "Months after invoice date"),
            ("months_after_end_of_month", "Months after end of month"),
        ]
        if payment_terms_delay_type:
            if payment_terms_delay_type == "months":
                date_type += months_delay_type
            elif payment_terms_delay_type == "weeks":
                date_type += weeks_delay_type
            elif payment_terms_delay_type == "weeks_and_months":
                date_type += weeks_delay_type + months_delay_type
        return date_type
```
I believe this helps improve the code readability for this section. WDYT?

File: account_payment_term_extension/views/res_config_settings_view.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        <field name="arch" type="xml">
+            <xpath expr="//app[@name='account']" position="inside">
+                <block title="Payment Terms" name="payment_terms_setting_container">
+                    <setting
+                        id="payment_terms_delay"
+                        string="Payment Terms Delay Types"
+                        company_dependent="1"
+                        help=""
+                        title="Configure the type of delay for payment terms."
+                    >
```
Comment (GuillemCForgeFlow) [inline account_payment_term_extension/views/res_config_settings_view.xml:19]: I'd say to add a little description for the purpose of the setting so it is visible for the user. Something like: "Select the Delay Type options that will be available in Payment Terms." What do you think?

File: account_payment_term_extension/models/res_company.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+# Copyright 2022 Moduon Team S.L.
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl).
```
Comment (GuillemCForgeFlow) [inline account_payment_term_extension/models/res_company.py]:
```suggestion
# Copyright 2024 ForgeFlow S.L. (https://www.forgeflow.com)
# License AGPL-3.0 or later (https://www.gnu.org/licenses/lgpl).
```
You can change the Copyright here.

File: account_payment_term_extension/views/res_config_settings_view.xml [hunk 4]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<!-- Copyright 2022 Moduon Team S.L.
+     License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl). -->
```
Comment (GuillemCForgeFlow) [inline account_payment_term_extension/views/res_config_settings_view.xml]: change the copyright here as well.

File: account_payment_term_extension/wizards/res_config_settings.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# Copyright 2024.
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl).
```
Comment (GuillemCForgeFlow) [inline account_payment_term_extension/wizards/res_config_settings.py]:
```suggestion
# Copyright 2024 ForgeFlow S.L. (https://www.forgeflow.com)
# License AGPL-3.0 or later (https://www.gnu.org/licenses/lgpl).
```
Also here.

**Review by GuillemCForgeFlow — CHANGES_REQUESTED** [review]
For the rest, functional and code review LGTM! 👍🏿

**Review by GuillemCForgeFlow — APPROVED** [review]
LGTM! Functional and code review 👍🏿

**Review by JaumeBforgeFlow — APPROVED** [review]

File: account_payment_term_extension/wizards/res_config_settings.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# Copyright 2024 ForgeFlow S.L. (https://www.forgeflow.com)
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/lgpl).
+
+from odoo import fields, models
+
+
+class ResConfigSettings(models.TransientModel):
```
Comment (Tisho99) [inline account_payment_term_extension/wizards/res_config_settings.py:7]: Shouldn't this be in the models folder?

**Review by Tisho99 — COMMENTED** [review]
This PR looks nice!

I have some suggestions, and them this would be approved for me

In addition, i got an error trying to configure the "Percent (untaxed amount)" payment value in lines. Steps to reproduce:
- Create a new payment term without saving it
- Change the value of the default payment term line to  "Percent (untaxed amount)"
- Add a new payment term line

File: account_payment_term_extension/readme/CONFIGURE.md [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+Term Lines, you need to:
+
+1.  Go to the menu Invoicing \> Configuration \> Invoicing \> Payment
+    Terms.
+
+To use multiple payment days, define for each payment term line which
+payment days apply, separated by spaces, commas or dashes. To use
+holidays, insert the holiday and the date payment terms will be
+postponed to.
+
```
Comment (Tisho99) [inline account_payment_term_extension/readme/CONFIGURE.md:11]: I think an explanation of how to configure Percent (Untaxed amount) should be added

File: account_payment_term_extension/models/res_company.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+    payment_terms_delay_type = fields.Selection(
+        string="Payment terms Delay types",
+        required=True,
+        default="days",
+        selection=[
+            ("days", "Days"),
+            ("weeks", "Days and Weeks"),
+            ("months", "Days and Months"),
+            ("weeks_and_months", "Days, Weeks and Months"),
```
Comment (Tisho99) [inline account_payment_term_extension/models/res_company.py:18]: Why weeks_and_months is not the default? In v16 you dont have to edit the settings to use this feature

File: account_payment_term_extension/wizards/res_config_settings.py (hunk 6 above)
Comment (AnaisLForgeFlow) [reply account_payment_term_extension/wizards/res_config_settings.py:7]: According to the Odoo Coding Guidelines, it is recommended to regroup the TransientModel classes and views in the wizard directory. Having that in mind, i should move the res_config_settings_view.xml file from the views folder to the wizards folder.
[https://www.odoo.com/documentation/17.0/contributing/development/coding_guidelines.html?highlight=file%20structure#module-structure](url)

File: account_payment_term_extension/models/res_company.py (hunk 8 above)
Comment (AnaisLForgeFlow) [reply account_payment_term_extension/models/res_company.py:18]: In previous versions like v16, Odoo included months and weeks in addition to days by default. However, in the v17, the default has been simplified to only include days. Given that the module extends the base functionality, the days are chosen as the default setting to keep consistency with the core functionality.

**Comment by GuillemCForgeFlow** [comment]
> This PR looks nice!
> 
> I have some suggestions, and them this would be approved for me
> 
> In addition, i got an error trying to configure the "Percent (untaxed amount)" payment value in lines. Steps to reproduce:
> 
> * Create a new payment term without saving it
> * Change the value of the default payment term line to  "Percent (untaxed amount)"
> * Add a new payment term line

I have tested based on the steps and didn't get the error. When you describe 'Change the value of the default payment term line...' you mean to just change the value from the variable `value` I assume?
Could you specify in more detail when exactly you are getting the error?

**Comment by Tisho99** [comment]
@GuillemCForgeFlow 

Yes, the steps are as you assumed.

And you are right, the error had dissapeared. However, i have updated account and account_payment_term_extension modules, and the error is back.

Anyway, with further debug i have found that the origin of the error is a change in the account module, so this PR is okay

**Review by Tisho99 — APPROVED** [review]
LGTM functional and technical review

**Review by HaraldPanten — APPROVED** [review]

**Pull request merged**